File: book/sum_range.js
// List representation
const Nil = { $: "Nil" };
const Cons = (head, tail) => ({ $: "Cons", head, tail });

function range(n, xs) {
  let result = xs;
  for (let i = n; i > 0; i--) {
    result = Cons(i - 1, result);
  }
  return result;
}

function sum(list) {
  let sum = 0;
  while (list.$ !== "Nil") {
    sum  = (sum + list.head) >>> 0;
    list = list.tail;
  }
  return sum;
}

function main() {
  console.log(sum(range(50_000_000, Nil)));
}

main();

File: book/sum_range.hs
import Data.Word

data List = Nil | Cons !Word32 List

range :: Word32 -> List -> List
range 0 xs = xs
range n xs = range (n - 1) (Cons (n - 1) xs)

sum' :: List -> Word32 -> Word32
sum' Nil              r = r
sum' (Cons head tail) r = sum' tail (head + r)

main :: IO ()
main = print $ sum' (range 50_000_000 Nil) 0
